src/chainstate/stacks/index/file.rs

```diff
@@ -0,0 +1,480 @@
+// Copyright (C) 2020 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::char::from_digit;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::convert::{TryFrom, TryInto};
+use std::env;
+use std::fmt;
+use std::fs;
+use std::fs::OpenOptions;
+use std::hash::{Hash, Hasher};
+use std::io;
+use std::io::{BufWriter, Cursor, Read, Seek, SeekFrom, Write};
+use std::iter::FromIterator;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::os;
+use std::path::{Path, PathBuf};
+use std::time::SystemTime;
+use std::{cmp, error};
+
+use rusqlite::{
+    types::{FromSql, ToSql},
+    Connection, Error as SqliteError, ErrorCode as SqliteErrorCode, OpenFlags, OptionalExtension,
+    Transaction, NO_PARAMS,
+};
+
+use chainstate::stacks::index::bits::{
+    get_node_byte_len, get_node_hash, read_block_identifier, read_hash_bytes, read_node_hash_bytes,
+    read_nodetype, read_nodetype_at_head, read_root_hash, write_nodetype_bytes,
+};
+use chainstate::stacks::index::node::{
+    clear_backptr, is_backptr, set_backptr, TrieNode, TrieNode16, TrieNode256, TrieNode4,
+    TrieNode48, TrieNodeID, TrieNodeType, TriePath, TriePtr,
+};
+use chainstate::stacks::index::storage::NodeHashReader;
+use chainstate::stacks::index::storage::TrieStorageConnection;
+use chainstate::stacks::index::Error;
+use chainstate::stacks::index::TrieLeaf;
+use chainstate::stacks::index::{trie_sql, ClarityMarfTrieId, MarfTrieId};
+
+use util_lib::db::sql_pragma;
+use util_lib::db::sqlite_open;
+use util_lib::db::tx_begin_immediate;
+use util_lib::db::tx_busy_handler;
+use util_lib::db::Error as db_error;
+use util_lib::db::SQLITE_MMAP_SIZE;
+
+use stacks_common::types::chainstate::BlockHeaderHash;
+use stacks_common::types::chainstate::BLOCK_HEADER_HASH_ENCODED_SIZE;
+use stacks_common::types::chainstate::{TrieHash, TRIEHASH_ENCODED_SIZE};
+
+/// Mapping between block IDs and trie offsets
+pub type TrieIdOffsets = HashMap<u32, u64>;
+
+/// Handle to a flat file containing Trie blobs
+pub struct TrieFileDisk {
+    fd: fs::File,
+    path: String,
+    trie_offsets: TrieIdOffsets,
+}
+
+/// Handle to a flat in-memory buffer containing Trie blobs (used for testing)
+pub struct TrieFileRAM {
+    fd: Cursor<Vec<u8>>,
+    readonly: bool,
+    trie_offsets: TrieIdOffsets,
+}
+
+pub enum TrieFile {
```

Comment:
Does a TrieFile contain multiple tries? I think a rustdoc for the `TrieFile` would be useful that explained what exactly a `TrieFile` contains (does it also store metadata?), and some brief explanation of how it is used by the rest of the marf implemenation.

Reply:
Yes -- a `TrieFile` is a flat storage space for tries (as opposed to a sqlite DB).  Will add rustdoc.